skxray/core/correlation.py

```diff
@@ -108,243 +250,104 @@ def multi_tau_auto_corr(num_levels, num_bufs, labels, images):
-    img_per_level = np.zeros(num_levels, dtype=np.int64)
-
-    start_time = time.time()  # used to log the computation time (optionally)
-
-    for n, img in enumerate(images):
+        raise ValueError("There must be an even number of `num_bufs`. You "
+                         "provided %s" % num_bufs)
+    if internal_state is None:
+        internal_state = _init_state(num_levels, num_bufs, labels)
+    # create a shorthand reference to the results and state named tuple
+    s = internal_state
+    # stash the number of pixels in the mask
+    num_pixels = np.bincount(s.label_mask)
+    # Convert from num_levels, num_bufs to lag frames.
+    tot_channels, lag_steps = multi_tau_lags(num_levels, num_bufs)
 
-        cur[0] = (1 + cur[0]) % num_bufs  # increment buffer
+    # iterate over the images to compute multi-tau correlation
+    for image in image_iterable:
+        # Compute the correlations for all higher levels.
+        level = 0
 
-        # Put the image into the ring buffer.
-        buf[0, cur[0] - 1] = (np.ravel(img))[pixel_list]
+        # increment buffer
+        s.cur[0] = (1 + s.cur[0]) % num_bufs
 
+        # Put the ROI pixels into the ring buffer.
+        s.buf[0, s.cur[0] - 1] = np.ravel(image)[s.pixel_list]
+        buf_no = s.cur[0] - 1
         # Compute the correlations between the first level
         # (undownsampled) frames. This modifies G,
-        # past_intensity_norm, future_intensity_norm,
+        # past_intensity, future_intensity,
         # and img_per_level in place!
-        _process(buf, G, past_intensity_norm,
-                 future_intensity_norm, label_mask,
-                 num_bufs, num_pixels, img_per_level,
-                 level=0, buf_no=cur[0] - 1)
+        _one_time_process(s.buf, s.G, s.past_intensity, s.future_intensity,
+                          s.label_mask, num_bufs, num_pixels, s.img_per_level,
+                          level, buf_no)
 
         # check whether the number of levels is one, otherwise
         # continue processing the next level
         processing = num_levels > 1
 
-        # Compute the correlations for all higher levels.
         level = 1
         while processing:
-            if not track_level[level]:
-                track_level[level] = 1
+            if not s.track_level[level]:
+                s.track_level[level] = True
                 processing = False
             else:
-                prev = 1 + (cur[level - 1] - 2) % num_bufs
-                cur[level] = 1 + cur[level] % num_bufs
+                prev = (1 + (s.cur[level - 1] - 2) % num_bufs)
+                s.cur[level] = (
+                    1 + s.cur[level] % num_bufs)
 
-                buf[level, cur[level] - 1] = (buf[level - 1, prev - 1] +
-                                              buf[level - 1,
-                                                  cur[level - 1] - 1])/2
+                # TODO clean this up. it is hard to understand
```

Comment:
@ericdill here the frames add average according to the multi tau levels and buffers